app/src/main/java/org/simple/clinic/teleconsultlog/prescription/medicines/TeleconsultMedicinesView.kt

```diff
@@ -0,0 +1,211 @@
+        defaultFrequency = teleconsultMedicinesConfig.defaultFrequency
+    ))
+  }
+
+  override fun showNoMedicines() {
+    emptyMedicinesTextView.visibility = VISIBLE
+    medicinesRecyclerView.visibility = GONE
+  }
+
+  override fun openEditMedicines(patientUuid: UUID) {
+    screenRouter.push(PrescribedDrugsScreenKey(patientUuid))
+  }
+
+  override fun openDrugDurationSheet(prescription: PrescribedDrug) {
+    val durationInDays = prescription.durationInDays?.toString() ?: teleconsultMedicinesConfig.defaultDuration.toDays().toString()
+    val intent = DrugDurationSheet.intent(
+        context = context,
+        drugDuration = DrugDuration(
+            uuid = prescription.uuid,
+            name = prescription.name,
+            dosage = prescription.dosage,
+            duration = durationInDays
+        )
+    )
+    activity.startActivityForResult(intent, DRUG_DURATION_SHEET)
+  }
+
+  override fun openDrugFrequencySheet(prescription: PrescribedDrug) {
+    val frequency = prescription.frequency ?: teleconsultMedicinesConfig.defaultFrequency
+    val intent = MedicineFrequencySheet.intent(
+        context = context,
+        medicineFrequencySheetExtra = MedicineFrequencySheetExtra(
+            uuid = prescription.uuid,
+            name = prescription.name,
+            dosage = prescription.dosage,
+            medicineFrequency = frequency
+        )
+    )
+    activity.startActivityForResult(intent, DRUG_FREQUENCY_SHEET)
+  }
+
+  private fun editClicks(): Observable<UiEvent> {
+    return medicinesEditButton
+        .clicks()
+        .map { EditMedicinesClicked }
+  }
+
+  private fun drugDurationChanges(): Observable<UiEvent> {
+    return screenRouter.streamScreenResults()
+        .ofType<ActivityResult>()
+        .extractSuccessful(DRUG_DURATION_SHEET) { intent ->
+          DrugDurationSheet.readSavedDrugDuration(intent)
+        }
+        .map { (uuid, duration) -> DrugDurationChanged(uuid, duration) }
+  }
+
+  private fun drugFrequencyChanges(): Observable<UiEvent> {
+    return screenRouter.streamScreenResults()
+        .ofType<ActivityResult>()
+        .extractSuccessful(DRUG_FREQUENCY_SHEET) { intent ->
+          MedicineFrequencySheet.readSavedDrugFrequency(intent)
+        }
+        .map { (uuid, frequency) -> DrugFrequencyChanged(uuid, frequency) }
+  }
+
+  private fun drugDurationClicks(): Observable<UiEvent> {
+    return teleconsultMedicinesAdapter.itemEvents
+        .filter { it is DrugDurationButtonClicked }
+        .map {
+          val drugDuration = it as DrugDurationButtonClicked
+          DrugDurationClicked(drugDuration.prescribedDrug)
+        }
+  }
+
+  private fun drugFrequencyClicks(): Observable<UiEvent> {
+    return teleconsultMedicinesAdapter.itemEvents
+        .filter { it is DrugFrequencyButtonClicked }
+        .map {
+          val drugDuration = it as DrugFrequencyButtonClicked
+          DrugFrequencyClicked(drugDuration.prescribedDrug)
+        }
+  }
```

Comment:
One small suggestion would be to use the `.ofType<>` operator which will allow us to skip doing a filter and a forced cast again.

This is something we can do in a later PR.

Reply:
Ah okay. TIL